pallets/parachain-staking/src/lib.rs

```diff
@@ -834,34 +841,182 @@ pub mod pallet {
 		/// - Writes: CollatorState, CandidatePool, ExitQueue, N * AtStake,
 		///   SelectedCandidates
 		/// # </weight>
-		#[pallet::weight(<T as pallet::Config>::WeightInfo::leave_candidates(T::MaxCollatorCandidates::get(), T::MaxCollatorCandidates::get() * T::MaxDelegatorsPerCollator::get()))]
-		pub fn leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::init_leave_candidates(
+			T::MaxCollatorCandidates::get(),
```

Comment:
Why don't we use the actual number of collators currently active in the set?

Reply:
I mean, here we are just always considering the worst case?

Reply:
The weight calculation shouldn't include any storage calls, because calculating the weight needs to be super lightweight and fast. So there are multiple different approaches to calculate the weight.

* the preferred way is to calculate the exact weight from the call parameters
* if the parameters don't provide enough info for the weight, assume the worst case and pay fees back after we know the exact weight
* if the worst case is too heavy and extreme (uses the whole block or similar), the user provides the values for calculating the weight and those values are checked first in the call.